token/core/zkatdlog/nogh/v1/crypto/upgrade/service.go

```diff
@@ -0,0 +1,249 @@
+)
+
+const (
+	ChallengeSize = 32
+)
+
+type (
+	Signature = []byte
+)
+
+//go:generate counterfeiter -o mock/des.go -fake-name Deserializer . Deserializer
+
+type Deserializer interface {
+	GetOwnerVerifier(id driver.Identity) (driver.Verifier, error)
+}
+
+//go:generate counterfeiter -o mock/ip.go -fake-name IdentityProvider . IdentityProvider
+
+type IdentityProvider interface {
+	GetSigner(id driver.Identity) (driver.Signer, error)
+}
+
+type Service struct {
+	Logger                          logging.Logger
+	MaxPrecision                    uint64
+	UpgradeSupportedTokenFormatList []token.Format
+	Deserializer                    Deserializer
+	IdentityProvider                IdentityProvider
+}
+
+func NewService(
+	logger logging.Logger,
+	maxPrecision uint64,
+	deserializer Deserializer,
+	identityProvider IdentityProvider,
+) (*Service, error) {
+	// compute supported tokens
+	var upgradeSupportedTokenFormatList []token.Format
+	for _, precision := range []uint64{16, 32, 64} {
+		format, err := v1.SupportedTokenFormat(precision)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed computing fabtoken token format with precision [%d]", precision)
+		}
+		if precision > maxPrecision {
+			upgradeSupportedTokenFormatList = append(upgradeSupportedTokenFormatList, format)
+		}
+	}
+
+	return &Service{
+		Logger:                          logger,
+		MaxPrecision:                    maxPrecision,
+		UpgradeSupportedTokenFormatList: upgradeSupportedTokenFormatList,
+		Deserializer:                    deserializer,
+		IdentityProvider:                identityProvider,
+	}, nil
+}
+
+func (s *Service) NewUpgradeChallenge() (driver.TokensUpgradeChallenge, error) {
+	// generate a 32 bytes secure random slice
+	key := make([]byte, ChallengeSize)
+	_, err := rand.Read(key)
+	if err != nil {
+		return nil, errors.Wrap(err, "error getting random bytes")
+	}
+	// rand.Read guarantees that len(key) == ChallengeSize, let's check it anyway
+	if len(key) != ChallengeSize {
+		return nil, errors.Errorf("invalid key size, got only [%d], expected [%d]", len(key), ChallengeSize)
+	}
+	return key, nil
+}
+
+// GenUpgradeProof does the following: For each token in input, it signs the concatenation of the challenge and the tokens to be upgraded.
+// These signatures are then added to the proof
+func (s *Service) GenUpgradeProof(ch driver.TokensUpgradeChallenge, ledgerTokens []token.LedgerToken, witness driver.TokensUpgradeWitness) (driver.TokensUpgradeProof, error) {
+	if len(ch) != ChallengeSize {
+		return nil, errors.Errorf("invalid challenge size, got [%d], expected [%d]", len(ch), ChallengeSize)
+	}
+	if len(ledgerTokens) == 0 {
+		return nil, errors.Errorf("no ledger tokens provided")
+	}
+	if len(witness) != 0 {
+		return nil, errors.Errorf("proof witness not expected")
+	}
+
+	proof := &Proof{
+		Challenge:  ch,
+		Tokens:     ledgerTokens,
+		Signatures: make([]Signature, 0, len(ledgerTokens)),
+	}
+	digest, err := proof.SHA256Digest()
```

Comment:
I would make a static method `SHA256Digest(challenge, tokens)` instead of defining it on the proof struct. The reason is because its normal use is based on an incomplete struct. If a struct doesn't have complete data, it should not be used and we should not call methods upon it, because this is like exposing its internal workings.

Instead, for code clarity, I would aim for as much immutability as possible: When we instantiate a struct we don't try to modify its content. Or at least we don't interact with it until it is fully instantiated.

Reply:
okay